{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HMM classification - part 2\n",
    "\n",
    "This doc lives in:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/alessandro/Unibo/python-programming-alessandro-lussana/LB1/prj_hmm_classification/docs\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the previous episode... workflow completed until the generation of fasta file with choosen 25 kunitz-containing sequences (after clustering and selection of representative sequences to avoid redundancies)\n",
    "\n",
    "## Model generation in dumb way (sequence only)\n",
    "\n",
    "Will be worth to demonstrate that taking into account the structural alignment allows to increase the alignment correctess, and consequently the classification performance of the resulting model\n",
    "\n",
    "### Step1: run muscle for MSAing the 25 kunitz seqs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/alessandro/Unibo/python-programming-alessandro-lussana/LB1/prj_hmm_classification/dataset\n",
      ">1BUN_B\n",
      "--------------------------RKRHPDCDK-PPDTKICQTVVRAFYYKPSAKRCV\n",
      "QFRYGGCNGNGNHFKSDHLCRCECLEYR------------\n",
      ">4NTW_B\n",
      "--------------------------EIRPAFCYEDPPFFQKCGAFVDSYYFNRSRITCV\n",
      "HFFYGQCDVNQNHFTTMSECNRVCHG--------------\n",
      ">1DTX_A\n",
      "--------------------------EPRRKLCIL-HRNPGRCYDKIPAFYYNQKKKQCE\n",
      "RFDWSGCGGNSNRFKTIEECRRTCIG--------------\n",
      ">5YV7_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Provided cores: 1\n",
      "Rules claiming more threads will be scaled down.\n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tmuscle_align\n",
      "\t1\n",
      "\n",
      "rule muscle_align:\n",
      "    input: chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.fasta.gz\n",
      "    output: chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
      "    jobid: 0\n",
      "    wildcards: sequences=chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95\n",
      "\n",
      "muscle -in <(zcat chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.fasta.gz) > chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "63 25 seqs, lengths min 54, max 86, avg 62\n",
      "00:00:00    19 MB(-3%)  Iter   1    0.31%  K-mer dist pass 1\r",
      "00:00:00    19 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\r\n",
      "00:00:00    19 MB(-3%)  Iter   1    0.31%  K-mer dist pass 2\r",
      "00:00:00    19 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\r\n",
      "00:00:00    19 MB(-3%)  Iter   1    4.17%  Align node       \r",
      "00:00:00    21 MB(-3%)  Iter   1    8.33%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   12.50%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   16.67%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   20.83%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   25.00%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   29.17%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   33.33%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   37.50%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   41.67%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   45.83%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   50.00%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   54.17%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   58.33%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   62.50%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   66.67%  Align node\r",
      "00:00:00    21 MB(-3%)  Iter   1   70.83%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   75.00%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   79.17%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   83.33%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   87.50%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   91.67%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1   95.83%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1  100.00%  Align node\r",
      "00:00:00    22 MB(-3%)  Iter   1  100.00%  Align node\r\n",
      "00:00:00    22 MB(-3%)  Iter   1    4.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1    8.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   12.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   16.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   20.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   24.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   28.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   32.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   36.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   40.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   44.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   48.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   52.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   56.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   60.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   64.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   68.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   72.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   76.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   80.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   84.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   88.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   92.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1   96.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1  100.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   1  100.00%  Root alignment\r\n",
      "00:00:00    22 MB(-3%)  Iter   2    4.35%  Refine tree   \r",
      "00:00:00    22 MB(-3%)  Iter   2    8.70%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   13.04%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   17.39%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   21.74%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   26.09%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   30.43%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   34.78%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   39.13%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   43.48%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   47.83%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   52.17%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   56.52%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   60.87%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   65.22%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   69.57%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2   73.91%  Refine tree\r",
      "00:00:00    22 MB(-3%)  Iter   2  100.00%  Refine tree\r\n",
      "00:00:00    22 MB(-3%)  Iter   2    4.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2    8.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   12.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   16.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   20.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   24.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   28.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   32.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   36.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   40.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   44.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   48.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   52.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   56.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   60.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   64.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   68.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   72.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   76.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   80.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   84.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   88.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   92.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2   96.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2  100.00%  Root alignment\r",
      "00:00:00    22 MB(-3%)  Iter   2  100.00%  Root alignment\r\n",
      "00:00:00    22 MB(-3%)  Iter   2  100.00%  Root alignment\r\n",
      "00:00:00    22 MB(-3%)  Iter   3    4.26%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3    6.38%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3    8.51%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   10.64%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   12.77%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   14.89%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   17.02%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   19.15%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   21.28%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   23.40%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   25.53%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   27.66%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   29.79%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   31.91%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   34.04%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   36.17%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   38.30%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   40.43%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   42.55%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   44.68%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   46.81%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   48.94%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   51.06%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   53.19%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   55.32%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   57.45%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   59.57%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   61.70%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   63.83%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   65.96%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   68.09%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   70.21%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   72.34%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   74.47%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   76.60%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   78.72%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   80.85%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   82.98%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   85.11%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   87.23%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   89.36%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   91.49%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   93.62%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   95.74%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3   97.87%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3  100.00%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3  102.13%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   3  100.00%  Refine biparts\r\n",
      "00:00:00    22 MB(-3%)  Iter   4    4.26%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4    6.38%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4    8.51%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   10.64%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   12.77%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   14.89%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   17.02%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   19.15%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   21.28%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   23.40%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   25.53%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   27.66%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   29.79%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   31.91%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   34.04%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   36.17%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   38.30%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   40.43%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   42.55%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   44.68%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   46.81%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   48.94%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   51.06%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   53.19%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   55.32%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   57.45%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   59.57%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   61.70%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   63.83%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   65.96%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   68.09%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   70.21%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   72.34%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   74.47%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   76.60%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   78.72%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   80.85%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   82.98%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   85.11%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   87.23%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   89.36%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   91.49%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   93.62%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   95.74%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4   97.87%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4  100.00%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4  102.13%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   4  100.00%  Refine biparts\r\n",
      "00:00:00    22 MB(-3%)  Iter   5    4.26%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   5  100.00%  Refine biparts\r\n",
      "00:00:00    22 MB(-3%)  Iter   5  100.00%  Refine biparts\r\n",
      "00:00:00    22 MB(-3%)  Iter   6    4.26%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6    6.38%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6    8.51%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   10.64%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   12.77%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   14.89%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   17.02%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   19.15%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   21.28%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   23.40%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   25.53%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   27.66%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   29.79%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   31.91%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   34.04%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   36.17%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   38.30%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   40.43%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   42.55%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   44.68%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   46.81%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   48.94%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   51.06%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   53.19%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   55.32%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   57.45%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   59.57%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   61.70%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   63.83%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   65.96%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   68.09%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   70.21%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   72.34%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   74.47%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   76.60%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   78.72%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   80.85%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   82.98%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   85.11%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   87.23%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   89.36%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   91.49%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   93.62%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   95.74%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6   97.87%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6  100.00%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6  102.13%  Refine biparts\r",
      "00:00:00    22 MB(-3%)  Iter   6  100.00%  Refine biparts\r\n",
      "Finished job 0.\n",
      "1 of 1 steps (100%) done\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../dataset/\n",
    "pwd\n",
    "snakemake -p chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
    "head chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step2: run hmmbuild to build the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/alessandro/Unibo/python-programming-alessandro-lussana/LB1/prj_hmm_classification/dataset\n",
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
      "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the GNU General Public License (GPLv3).\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
      "# output HMM file:                  chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle    25   100    60     2.07  0.931 \n",
      "\n",
      "# CPU time: 0.06u 0.00s 00:00:00.06 Elapsed: 00:00:00.07\n",
      "HMMER3/f [3.1b2 | February 2015]\n",
      "NAME  chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle\n",
      "LENG  60\n",
      "ALPH  amino\n",
      "RF    no\n",
      "MM    no\n",
      "CONS  yes\n",
      "CS    no\n",
      "MAP   yes\n",
      "DATE  Thu Apr  4 12:08:08 2019\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Provided cores: 1\n",
      "Rules claiming more threads will be scaled down.\n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tbuild_hmm\n",
      "\t1\n",
      "\n",
      "rule build_hmm:\n",
      "    input: chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
      "    output: chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.hmm\n",
      "    jobid: 0\n",
      "    wildcards: msa=chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95, source=muscle\n",
      "\n",
      "hmmbuild --informat afa chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.hmm chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.afa\n",
      "Finished job 0.\n",
      "1 of 1 steps (100%) done\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ../dataset/\n",
    "pwd\n",
    "snakemake -p chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.hmm\n",
    "head chosen_blastclust_seq_for_intersec_rcsbpdb_wt_PF00014_len50_90_and_pdbefold_5pti_A_summary_S99_L0.95.muscle.hmm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\[classification and assessment of performance will be done later\\]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
